sklearn/utils/__init__.py

```diff
@@ -215,10 +219,19 @@ def _determine_key_type(key, accept_slice=True):
             raise ValueError(err_msg)
         return key_type.pop()
     if hasattr(key, "dtype"):
-        try:
-            return array_dtype_to_str[key.dtype.kind]
-        except KeyError:
-            raise ValueError(err_msg)
+        xp, is_array_api = get_namespace(key)
+        if is_array_api and not _is_numpy_namespace(xp):
```

Comment:
```suggestion
        # NumPy arrays are special-cased in their own branch because the Array API
        # cannot handle string-based dtypes.
        if is_array_api and not _is_numpy_namespace(xp):
```